### PR TITLE
[keyboard] add new keyboard: keyboard/willow

### PR DESCRIPTION
**Pull request by hanachi-ap**
## Description
Add new keyboard,  willow/willow64.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/willow/willow64/config.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (drashna) [inline keyboards/willow/willow64/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/willow/willow64/config.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#pragma once
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID 0xBB01
+#define PRODUCT_ID 0x0001
+#define DEVICE_VER 0x0002
+#define MANUFACTURER Hanachi
+#define PRODUCT Willow64
+#define DESCRIPTION Willow64 ergo keyboard
```
Comment (drashna) [inline keyboards/willow/willow64/config.h:13]:
```suggestion
```

File: keyboards/willow/willow64/config.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+#define MATRIX_ROWS 10
+#define MATRIX_COLS 8
+
+#define MATRIX_ROW_PINS { D4, C6, D7, E6, B4}
+#define MATRIX_COL_PINS { F4, F5, F6, F7, B1, B3, B2, B6}
+#define UNUSED_PINS
+#define DIODE_DIRECTION COL2ROW
+
+#define MASTER_LEFT
```
Comment (drashna) [inline keyboards/willow/willow64/config.h:23]: Not actually needed, as this is the default behavior.
```suggestion
```

File: keyboards/willow/willow64/config.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#define SOFT_SERIAL_PIN D2
+#define USE_SERIAL
```
Comment (drashna) [inline keyboards/willow/willow64/config.h:25]: Also, default behavior. 
```suggestion
```

File: keyboards/willow/willow64/keymaps/buildtest/keymap.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+#define DUMMY KC_SPC
+
+#define TESTLAYOUT( \
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/buildtest/keymap.c:21]:
```suggestion
#define LAYOUT_test( \
```

File: keyboards/willow/willow64/keymaps/buildtest/keymap.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        { DUMMY, R01, R02, R03, R04, R05, R06 ,R07  }, \
+        { DUMMY, R08, R09, R10, R11, R12, R13, R14  }, \
+        { DUMMY, R15, R16 ,R17, R18, R19, R20, R21  }, \
+        { R22, R23, R24, R25, R26 ,R27, R28, R29  }, \
+        { R30, R31, R32, R33, DUMMY, DUMMY, R34, R35 }  \
+    }
+
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+[0] = TESTLAYOUT(
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/buildtest/keymap.c:43]:
```suggestion
[0] = LAYOUT_test(
```

File: keyboards/willow/willow64/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include QMK_KEYBOARD_H
+
+#define _QWERTY 0
+#define _FN1 1
+#define _FN2 2
+#define _FN3 3
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/default/keymap.c:22]: We highly recommend using enums for the layer naming as it is easier to manage long term. Especially as people will copy the default. 

```suggestion
enum custom_layers {
    _QWERTY,
    _FN1,
    _FN2,
    _FN3,
};
```

File: keyboards/willow/willow64/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+    QWERTY = SAFE_RANGE,
+};
+
+#define xxxx KC_NO
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/default/keymap.c:29]: These defines aren't necessary, as these are actually defined in the core.  You can safely remove them. 

https://github.com/qmk/qmk_firmware/blob/a0309db983150fd0197eb620a6ba552d90d29f93/quantum/quantum_keycodes.h#L30-L31


```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by hanachi-ap** [comment]
Thanks for the review.
I fixed the issues you pointed out.

**Review by drashna — APPROVED** [review]

**Review by tzarc — CHANGES_REQUESTED** [review]
Please add via support through a keymap, rather than defining a new keyboard variant for via.

File: keyboards/willow/willow64/keymaps/buildtest/config.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+#ifdef  RGBLED_SPLIT
+  #undef RGBLED_SPLIT
+  #define RGBLED_SPLIT {35, 35}
+#endif
+
+#ifdef RGBLED_NUM
+  #undef RGBLED_NUM
+  #define RGBLED_NUM 70
+#endif
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/buildtest/config.h:27]:
```suggestion
#ifdef  RGBLED_SPLIT
#    undef RGBLED_SPLIT
#    define RGBLED_SPLIT {35, 35}
#endif

#ifdef RGBLED_NUM
#    undef RGBLED_NUM
#    define RGBLED_NUM 70
#endif
```

File: keyboards/willow/willow64/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    _FN1,
+    _FN2,
+    _FN3,
+};
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+    QWERTY = SAFE_RANGE,
+};
+
```
Comment (noroadsleft) [inline keyboards/willow/willow64/keymaps/default/keymap.c:29]: This isn't being used.

```suggestion
```

File: keyboards/willow/willow64/keymaps/via/keymap.c [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    _FN1,
+    _FN2,
+    _FN3,
+};
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+    QWERTY = SAFE_RANGE,
+};
+
```
Comment (noroadsleft) [inline keyboards/willow/willow64/keymaps/via/keymap.c:29]:
```suggestion
```

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    _GAME,
+    _FN1,
+    _FN2,
+    _FN3,
+};
+
+enum custom_keycodes {
+    QWERTY = SAFE_RANGE,
+};
+
```
Comment (noroadsleft) [inline keyboards/willow/willow64/keymaps/hanachi/keymap.c:30]:
```suggestion
```

File: keyboards/willow/willow64/rules.mk [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (noroadsleft) [inline keyboards/willow/willow64/rules.mk:11]:
```suggestion
```

File: keyboards/willow/willow64/rules.mk [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+BOOTLOADER = caterina
+
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
```
Comment (noroadsleft) [inline keyboards/willow/willow64/rules.mk:16]:
```suggestion
#   change yes to no to disable
```

File: keyboards/willow/willow64/rules.mk [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+BACKLIGHT_ENABLE = no      # Enable keyboard backlight functionality
+MIDI_ENABLE = no           # MIDI controls
+AUDIO_ENABLE = no          # Audio output on port C6
+UNICODE_ENABLE = no        # Unicode
+BLUETOOTH_ENABLE = no      # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = yes      # Enable WS2812 RGB underlight.
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+OLED_DRIVER_ENABLE = no      # OLED display
+SPLIT_KEYBOARD = yes
```
Comment (noroadsleft) [inline keyboards/willow/willow64/rules.mk:33]: Formatting update and removing unnecessary rules.

```suggestion
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

SPLIT_KEYBOARD = yes
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
The changes to `lib/chibios` and `lib/chibios-contrib` also must be reverted.

File: keyboards/willow/willow64/readme.md [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+This is a 64 keys custom keyboard with a willow layout.
+
+- Keyboard Maintainer: [hanachi](https://github.com/hanachi-ap/) [@haswellep](https://twitter.com/haswellep)  
+- Hardware Supported:  Willow64 PCB, Pro Micro  
+- Hardware Availability: https://hanachiya.booth.pm/ 
+
+Make example for this keyboard (after setting up your build environment):
+
+    qmk compile -kb willow/willow64 -km default
```
Comment (fauxpark) [inline keyboards/willow/willow64/readme.md:13]:
```suggestion
    make willow/willow64:default
```

File: keyboards/willow/willow64/via/readme.md [hunk 17]
```diff
@@ -0,0 +1,6 @@
+
+# Build with VIA enabled
+
+Make example for this keyboard (with VIA enabled):
+
+    qmk compile -kb willow/willow64/via -km default
```
Comment (fauxpark) [inline keyboards/willow/willow64/via/readme.md:6]: This file can be deleted.

File: keyboards/willow/willow64/via/rules.mk [hunk 18]
```diff
@@ -0,0 +1,2 @@
+
+VIA_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/willow/willow64/via/rules.mk:2]: This file can be deleted.

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    _QWERTY,
+    _GAME,
+    _FN1,
+    _FN2,
+    _FN3,
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+[_QWERTY]=LAYOUT(   //QWERTY
+	KC_ESC	,	KC_1	,	KC_2	,	KC_3	,	KC_4	,	KC_5	,					KC_6	,	KC_7	,	KC_8	,	KC_9	,	KC_0	,	KC_MINS	,	KC_EQL	,	KC_GRV	,	KC_BSLS	,	\
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/hanachi/keymap.c:29]: The backslashes at the end of these lines are unnecessary.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c (hunk 19 above)
Comment (hanachi-ap) [reply keyboards/willow/willow64/keymaps/hanachi/keymap.c]: Removed unnecessary backslashes.

File: keyboards/willow/willow64/via/readme.md (hunk 17 above)
Comment (hanachi-ap) [reply keyboards/willow/willow64/via/readme.md]: Unnecessary files and directories have been removed.

File: keyboards/willow/willow64/keymaps/buildtest/config.h [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+#ifdef  RGBLED_SPLIT
+#    undef RGBLED_SPLIT
+#    define RGBLED_SPLIT {35, 35}
+#endif
+
+#ifdef RGBLED_NUM
+#    undef RGBLED_NUM
+#    define RGBLED_NUM 70
+#endif
```
Comment (drashna) [inline keyboards/willow/willow64/keymaps/buildtest/config.h:27]: Could just do this: 
```suggestion
#if defined(RGBLED_SPLIT) || defined(RGBLED_NUM)
#    undef RGBLED_SPLIT
#    undef RGBLED_NUM
#    define RGBLED_SPLIT { 35, 35 }
#    define RGBLED_NUM 70
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]
You have changes to the git submodules. 

You may need to rebase to fix this.

**Comment by hanachi-ap** [comment]
@drashna , thanks for the review.
I was a bit confused by your point, but I think the submodules are now in the right state.

File: keyboards/willow/willow64/willow64.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+        { L01, L02, L03, L04, L05, L06, NOP, NOP  }, \
+        { L07, L08, L09, L10, L11, L12, NOP, NOP  }, \
+        { L13, L14, L15, L16, L17, L18, NOP, NOP  }, \
+        { L19, L20, L21, L22, L23, L24, L25, NOP  }, \
+        { NOP, NOP, NOP, L26, L27, L28, L29, NOP  }, \
+        { NOP, R01, R02, R03, R04, R05, R06 ,R07  }, \
+        { NOP, R08, R09, R10, R11, R12, R13, R14  }, \
+        { NOP, R15, R16 ,R17, R18, R19, R20, R21  }, \
+        { R22, R23, R24, R25, R26 ,R27, R28, R29  }, \
+        { R30, R31, R32, R33, NOP, NOP, R34, R35 }  \
```
Comment (fauxpark) [inline keyboards/willow/willow64/willow64.h:38]:
```suggestion
#define XXX KC_NO

#define LAYOUT( \
    L01, L02, L03, L04, L05, L06,              R01, R02, R03, R04, R05, R06, R28, R07, R35, \
    L07, L08, L09, L10, L11, L12,              R08, R09, R10, R11, R12, R13, R34, R14, \
    L13, L14, L15, L16, L17, L18,              R15, R16, R17, R18, R19, R20, R21,  \
    L19, L20, L21, L22, L23, L24, L25,    R22, R23, R24, R25, R26, R27,        R29, \
                   L26, L27, L28, L29,    R30, R31, R32, R33                      \
    ) \
    { \
        { L01, L02, L03, L04, L05, L06, XXX, XXX  }, \
        { L07, L08, L09, L10, L11, L12, XXX, XXX  }, \
        { L13, L14, L15, L16, L17, L18, XXX, XXX  }, \
        { L19, L20, L21, L22, L23, L24, L25, XXX  }, \
        { XXX, XXX, XXX, L26, L27, L28, L29, XXX  }, \
        { XXX, R01, R02, R03, R04, R05, R06 ,R07  }, \
        { XXX, R08, R09, R10, R11, R12, R13, R14  }, \
        { XXX, R15, R16 ,R17, R18, R19, R20, R21  }, \
        { R22, R23, R24, R25, R26 ,R27, R28, R29  }, \
        { R30, R31, R32, R33, XXX, XXX, R34, R35 }  \
```

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+    debug_enable = true;
+    debug_matrix = true;
+#endif
+}
+
+#ifdef VIRTSER_ENABLE
+void virtser_recv(const uint8_t ch) {
+#    ifdef CONSOLE_ENABLE
+    uprintf("recv  %u\n", ch);
+#    endif
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/hanachi/keymap.c:129]:
```suggestion
void virtser_recv(const uint8_t ch) {
    uprintf("recv  %u\n", ch);
```

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+};
+
+int layer = _QWERTY;
+
+bool process_record_user(uint16_t keycode, keyrecord_t* record) {
+#ifdef CONSOLE_ENABLE
+    uprintf("KL: kc: %u, col: %u, row: %u, pressed: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed);
+#endif
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/hanachi/keymap.c:71]:
```suggestion
    uprintf("KL: kc: %u, col: %u, row: %u, pressed: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed);
```

File: keyboards/willow/willow64/keymaps/buildtest/keymap.c [hunk 24]
```diff
@@ -0,0 +1,58 @@
+/* Copyright 2021 Hanachi
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/buildtest/keymap.c:1]: What's the purpose of this keymap?

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/willow/willow64/keymaps/buildtest/keymap.c (hunk 24 above)
Comment (hanachi-ap) [reply keyboards/willow/willow64/keymaps/buildtest/keymap.c:1]: This is used to check the lighting of the LED when soldering.

**Review by drashna — APPROVED** [review]

File: keyboards/willow/willow64/readme.md [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+
+- Keyboard Maintainer: [hanachi](https://github.com/hanachi-ap/) [@haswellep](https://twitter.com/haswellep)  
+- Hardware Supported:  Willow64 PCB, Pro Micro  
+- Hardware Availability: https://hanachiya.booth.pm/ 
+
+Make example for this keyboard (after setting up your build environment):
+
+    make willow/willow64:default
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
Comment (noroadsleft) [inline keyboards/willow/willow64/readme.md:15]: Could you please add instructions above this line on how to reset the keyboard into bootloader mode?

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/willow/willow64/readme.md (hunk 25 above)
Comment (hanachi-ap) [reply keyboards/willow/willow64/readme.md:15]: Thanks for the review.
Is this what your suggestion is?
`To enter bootloader mode, press the reset button on the keyboard.`

Comment (noroadsleft) [reply keyboards/willow/willow64/readme.md:15]: Yes.

Comment (hanachi-ap) [reply keyboards/willow/willow64/readme.md:17]: Thank you. I fixed it.

File: keyboards/willow/willow64/keymaps/buildtest/keymap.c [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+[0] = LAYOUT_test(
+	KC_ESC	,	KC_1	,	KC_2	,	KC_3	,	KC_4	,	KC_5	,					KC_6	,	KC_7	,	KC_8	,	KC_9	,	KC_0	,	KC_MINS	,	KC_EQL	,	KC_GRV	,	KC_BSPC	,	\
+	KC_TAB	,	KC_Q	,	KC_W	,	KC_E	,	KC_R	,	KC_T	,					KC_Y	,	KC_U	,	KC_I	,	KC_O	,	KC_P	,	KC_LBRC	,	KC_RBRC	,			KC_BSLS	,	\
+	KC_LCTL	,	KC_A	,	KC_S	,	KC_D	,	KC_F	,	KC_G	,					KC_H	,	KC_J	,	KC_K	,	KC_L	,	KC_SCLN	,	KC_QUOT	,					KC_ENT	,	\
+	KC_LSFT	,	KC_Z	,	KC_X	,	KC_C	,	KC_V	,	KC_B	,	KC_DEL	,	KC_LGUI	,	KC_N	,	KC_M	,	KC_COMM	,	KC_DOT	,	KC_SLSH	,							KC_RSFT	,	\
+							KC_LEFT   ,	KC_RIGHT,	KC_SPC	,	KC_LCTL	,	KC_ENT	,	KC_BSPC	,	KC_DOWN,	KC_UP														\
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/buildtest/keymap.c:48]:
```suggestion
	KC_ESC	,	KC_1	,	KC_2	,	KC_3	,	KC_4	,	KC_5	,					KC_6	,	KC_7	,	KC_8	,	KC_9	,	KC_0	,	KC_MINS	,	KC_EQL	,	KC_GRV	,	KC_BSPC	,
	KC_TAB	,	KC_Q	,	KC_W	,	KC_E	,	KC_R	,	KC_T	,					KC_Y	,	KC_U	,	KC_I	,	KC_O	,	KC_P	,	KC_LBRC	,	KC_RBRC	,			KC_BSLS	,
	KC_LCTL	,	KC_A	,	KC_S	,	KC_D	,	KC_F	,	KC_G	,					KC_H	,	KC_J	,	KC_K	,	KC_L	,	KC_SCLN	,	KC_QUOT	,					KC_ENT	,
	KC_LSFT	,	KC_Z	,	KC_X	,	KC_C	,	KC_V	,	KC_B	,	KC_DEL	,	KC_LGUI	,	KC_N	,	KC_M	,	KC_COMM	,	KC_DOT	,	KC_SLSH	,							KC_RSFT	,
							KC_LEFT   ,	KC_RIGHT,	KC_SPC	,	KC_LCTL	,	KC_ENT	,	KC_BSPC	,	KC_DOWN,	KC_UP
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/willow/willow64/keymaps/buildtest/rules.mk [hunk 27]
```diff
@@ -0,0 +1,4 @@
+
+CONSOLE_ENABLE = yes       # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+RGBLIGHT_ENABLE = yes      # Enable WS2812 RGB underlight.
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/buildtest/rules.mk:4]:
```suggestion
CONSOLE_ENABLE = yes
```

File: keyboards/willow/willow64/info.json [hunk 28]
```diff
@@ -0,0 +1,77 @@
+{
+    "keyboard_name": "Willow64",
+    "url": "https://github.com/hanachi-ap/willow64-doc",
+    "maintainer": "hanachi-ap",
+    "width": 18.5,
+    "height": 5,
```
Comment (fauxpark) [inline keyboards/willow/willow64/info.json:6]:
```suggestion
```

File: keyboards/willow/willow64/keymaps/hanachi/keymap.c [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+    return state;
+}
+
+bool led_update_user(led_t led_state) {
+    if (layer == _FN1) {
+        rgblight_set_layer_state(2, IS_HOST_LED_ON(USB_LED_NUM_LOCK));
+    } else {
+        rgblight_set_layer_state(2, false);
+    }
+    rgblight_set_layer_state(1, IS_HOST_LED_ON(USB_LED_CAPS_LOCK));
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/hanachi/keymap.c:106]:
```suggestion
        rgblight_set_layer_state(2, host_keyboard_led_state().caps_lock);
    } else {
        rgblight_set_layer_state(2, false);
    }
    rgblight_set_layer_state(1, host_keyboard_led_state().caps_lock);
```

File: keyboards/willow/willow64/keymaps/hanachi/rules.mk [hunk 30]
```diff
@@ -0,0 +1,2 @@
+
+AUTO_SHIFT_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/hanachi/rules.mk:2]:
```suggestion
AUTO_SHIFT_ENABLE = yes
```

File: keyboards/willow/willow64/keymaps/via/rules.mk [hunk 31]
```diff
@@ -0,0 +1,2 @@
+
+VIA_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/willow/willow64/keymaps/via/rules.mk:2]:
```suggestion
VIA_ENABLE = yes
```

File: keyboards/willow/willow64/rules.mk [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+EXTRAKEY_ENABLE = no        # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
```
Comment (fauxpark) [inline keyboards/willow/willow64/rules.mk:21]:
```suggestion
```

File: keyboards/willow/willow64/rules.mk [hunk 33]
```diff
@@ -0,0 +1,24 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
```
Comment (fauxpark) [inline keyboards/willow/willow64/rules.mk:10]:
```suggestion
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/willow/willow64/willow64.h [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+        { L19, L20, L21, L22, L23, L24, L25, XXX  }, \
+        { XXX, XXX, XXX, L26, L27, L28, L29, XXX  }, \
+        { XXX, R01, R02, R03, R04, R05, R06 ,R07  }, \
+        { XXX, R08, R09, R10, R11, R12, R13, R14  }, \
+        { XXX, R15, R16 ,R17, R18, R19, R20, R21  }, \
+        { R22, R23, R24, R25, R26 ,R27, R28, R29  }, \
+        { R30, R31, R32, R33, XXX, XXX, R34, R35 }  \
+    }
+
+/* LED POSITION
```
[the remark this answers is not in the document]
Comment (hanachi-ap) [reply keyboards/willow/willow64/willow64.h:41]: Thanks, I will consider implementing the RGB matrix.

**Comment by tzarc** [comment]
This has two keyboards and core changes.
Please limit things to just the willow.

**Comment by hanachi-ap** [comment]
It seems that some unintended changes have entered my repository.

To clean up the repository, I will cancel this PR.
Thank you for all the feedback.

**Pull request closed without merging**